### PR TITLE
Assignment3 - Conner & Rahat

### PR DESCRIPTION
**Pull request by connerhanson**
Conner Hanson - @connerhanson
Rahat Bhatia - @rahat15

Locally it was working, but when trying to deploy it to azure it gives the same error as it did before I set the admin as we did in class 5/7.

https://agreeable-island-web-app.azurewebsites.net/         -- api url, still shows same error after setting admin as we did in class 5/7
https://agreeable-island-0e0c5251e.5.azurestaticapps.net/      -- shows leaderboard

### CODE REVIEW

File: Wordle.Api/Wordle.Api/Controllers/LeaderboardController.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      var top10Scores = _wordleService.GetTop10Scores();
+      return Ok(top10Scores);
+    }
+
+    [HttpPost]
+    public IActionResult PostScore(Player player)
+    {
+      _wordleService.AddOrUpdatePlayer(player);
+      return Ok();
+    }
```
Comment (SaaWest) [inline Wordle.Api/Wordle.Api/Controllers/LeaderboardController.cs]: One thing to think about is to make these return a task and also make these async. And for the GET you wanna make it return a task of  Ienumerable/List. One thing to consider for the GET is if you want it to return certain columns or all columns from the table and then just parse what you want into the table and return the top ten results. hint: `.Take(10)`

File: wordle-web/pages/leaderboard.vue [hunk 2]
```diff
@@ -0,0 +1,56 @@
+<template>
```
Comment (SaaWest) [inline wordle-web/pages/leaderboard.vue:1]: there is no way to navigate to this from the home page

File: wordle-web/pages/leaderboard.vue [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                    <th class="text-center text-h6 AverageAttempts">
+                        <strong>Average Guesses</strong>
+                    </th>
+                </tr>
+            </thead>
+            <tbody>
+                <tr v-for="player in topTenScores" :key="player.name">
+                    <td class="text-center">{{ player.name }}</td>
+                    <td class="text-center">{{ player.gameCount }}</td>
+                    <td class="text-center">{{ player.averageAttempts }}</td>
```
Comment (SaaWest) [inline wordle-web/pages/leaderboard.vue:22]: since you have the optional parameter for time you could implement the extra credit by `{{player.AverageSecondsPerGame}}`.

File: Wordle.Api/Wordle.Api/Services/WordleService.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+using System.Collections.Generic;
```
Comment (SaaWest) [inline Wordle.Api/Wordle.Api/Services/WordleService.cs:1]: it be a good idea to try and implement tests for this.

File: wordle-web/pages/leaderboard.vue [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import { ref } from 'vue';
+    import Axios from 'axios';
+
+const apiUrl = '';
+
+
+interface Player {
+  name: string;
+  gameCount: number;
+  averageAttempts: number;
```
Comment (SaaWest) [inline wordle-web/pages/leaderboard.vue]: for extra credit include the `AverageSecondsPerGame` in the interface

File: wordle-web/components/NameDialog.vue [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+</template>
+
+<script setup lang="ts">
+import {defineModel, inject} from 'vue-demi';
+import nuxtStorage from 'nuxt-storage';
+
+const modelValue = defineModel<boolean>({ default: false });
+const userName = inject("userName");
+
+function saveUsername() {
```
Comment (SaaWest) [inline wordle-web/components/NameDialog.vue:23]: nice

**Review by SaaWest — COMMENTED** [review]
## Instructions

**1. Create a leaderboard API Controller** ✅
- Create an endpoint that returns the top 10 scores (HttpGet) ✅
  - Results should include Name, average guesses, and number of games played ❌✅(its getting all columns so unable to determine the right course of action)
- Create an endpoint that allows for posting a score (HttpPost) with data in the body ✅
  - A new name should add a new record ✅
  - An existing name should update that record ✅
- Save and load all data with Entity Framework to an Azure SQL database or LocalDb for debugging ✅
  - Class should be named "Player" and "Players" for the DbSet ✅
  - Fields should be: 
    - int PlayerId ✅
    - string Name ✅
    - int GameCount ✅
    - double AverageAttempts ✅
    - int AverageSecondsPerGame ✅(OPTIONAL)
- The logic should reside in a service that is injected via dependency injection into the controller ✅
- Set up CORS (Cross Origin Resource Sharing) ✅

**2. Create a dialog for getting the user's name** ✅
- If we don't have the user's name, present the user with a dialog that allows them to type in their name ✅
- The users current name should show on the game page (upper right) ❌✅ (its on the Left)
  - This should be visible when setting/changing the name. It should change as the user types ✅
- The user should be able to click on their name to change it (with the same dialog) ✅
- Save the name in local storage and load it for the next game if it is set ✅
- If the name is not set, display Guest as the name. ✅
  - Clicking on this allows for preemptive setting of the name before the end of the game ✅
- Dialog must be persistent, doesn't close by clicking off the dialog ❌

**3. Save the Score and Seconds to complete** ✅
- Save the score and time to complete in seconds at the end of the game if we have a name. ✅
- If the user hasn't set a name, save the score and seconds as "Guest". (Be sure to prompt first) ✅
- Save should be done with an Axios Post ✅

**4. Create a Leaderboard Page** ✅
- Create a /leaderboard page in the application that shows the top scores ✅
- Use Axios to call the leaderboard API ✅
- Make a nice display of the returned data ❌✅(unable to determine the looks)
- Add a link on the front page to access the leaderboard ❌
- Add a link to the app bar to access the leaderboard ❌

**5. Unit testing** ❌(OPTIONAL) (Not covered in class, but in instructor's repo: https://github.com/granterickson/EWU-CSCD379-2024-Spring)
- Tests for the service ❌ (OPTIONAL)
- Integration tests for the controller ❌ (OPTIONAL)

**6. Deploy site to Azure** ✅
- Deploy API site with Github Action to Azure Web App ✅
- Unit Tests should run on build ❌


## Extra Credit

- Unit testing on the client side (3) ❌
- Animate the leaderboard (3) ❌
- Create a logo (3) ❌
- Add Average Seconds to win in score (3) ❌✅(just need it to appear in leaderboard)

**Comment by HadiM2** [comment]
Reviewing

File: .github/workflows/assignment3_agreeable-island-web-app.yml
Comment (HadiM2) [inline .github/workflows/assignment3_agreeable-island-web-app.yml:1]: Why did you change the Secret key?

**Review by HadiM2 — COMMENTED** [review]
# Instructions

1. **Create a leaderboard API Controller** ✅
    - Create an endpoint that returns the top 10 scores (HttpGet) ✅
        - Results should include Name, average guesses, and number of games played ❌ (Almost There)
    - Create an endpoint that allows for posting a score (HttpPost) with data in the body ✅
        - A new name should add a new record ✅
        - An existing name should update that record ✅
    - Save and load all data with Entity Framework to an Azure SQL database or LocalDb for debugging ✅
        - Class should be named "Player" and "Players" for the DbSet ✅
        - Fields should be:
            - int PlayerId ✅
            - string Name ✅
            - int GameCount ✅
            - double AverageAttempts ✅
            - int AverageSecondsPerGame ✅(OPTIONAL)
    - The logic should reside in a service that is injected via dependency injection into the controller ✅
    - Set up CORS (Cross Origin Resource Sharing) ✅

2. **Create a dialog for getting the user's name** ✅
    - If we don't have the user's name, present the user with a dialog that allows them to type in their name ✅
    - The user's current name should show on the game page (upper right) ✅ (Design choice, on the left)
        - This should be visible when setting/changing the name. It should change as the user types ✅
    - The user should be able to click on their name to change it (with the same dialog) ✅
    - Save the name in local storage and load it for the next game if it is set ✅
    - If the name is not set, display Guest as the name. ✅
        - Clicking on this allows for preemptive setting of the name before the end of the game ✅
    - Dialog must be persistent ❌

3. **Save the Score and Seconds to complete** ✅
    - Save the score and time to complete in seconds at the end of the game if we have a name. ✅
    - If the user hasn't set a name, save the score and seconds as "Guest". (Be sure to prompt first) ✅
    - Save should be done with an Axios Post ✅

4. **Create a Leaderboard Page** ✅
    - Create a /leaderboard page in the application that shows the top scores ✅
    - Use Axios to call the leaderboard API ✅
    - Make a nice display of the returned data ❌
    - Add a link on the front page to access the leaderboard ❌
    - Add a link to the app bar to access the leaderboard ❌

5. **Unit testing** ❌(OPTIONAL) (Not covered in class, but in instructor's repo: [GitHub Repo](https://github.com/granterickson/EWU-CSCD379-2024-Spring))
    - Tests for the service ❌ (OPTIONAL)
    - Integration tests for the controller ❌ (OPTIONAL)

6. **Deploy site to Azure** ✅
    - Deploy API site with Github Action to Azure Web App ✅
    - Unit Tests should run on build ❌

**No Extra Credit Done As Of Now**

File: Wordle.Api/Wordle.Api/Services/WordleService.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    {
+      _context = context;
+    }
+
+    public List<Player> GetTop10Scores()
+    {
+      return _context.Players.OrderByDescending(p => p.AverageAttempts).ThenBy(p => p.GameCount).Take(10).ToList();
+    }
+
+    public void AddOrUpdatePlayer(Player player)
```
Comment (montalcy) [inline Wordle.Api/Wordle.Api/Services/WordleService.cs]: I like how you made this into one whole method instead of two separate. Do you think making them separate would make for easier testing and code structure?

File: wordle-web/app.vue [hunk 8]
```diff
@@ -5,6 +5,8 @@
         <template v-slot:prepend>
           <v-icon color="secondary"> mdi-book </v-icon>
         </template>
+        <v-btn @click="showUserNameDialog = true">{{ displayedUserName }}</v-btn>
```
Comment (montalcy) [inline wordle-web/app.vue:8]: I dont see the username displayed? I could be wrong

File: wordle-web/components/NameDialog.vue [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  <v-dialog v-model="modelValue" max-width="500">
+    <v-card>
+      <v-sheet color="primary"></v-sheet>
+      <v-card-title>Enter a username.</v-card-title>
+      <v-text-field v-model="userName" label="UserName" required></v-text-field>
+      <v-card-actions>
+        <v-spacer></v-spacer>
+        <v-btn color="primary" @click="closeDialog">Cancel</v-btn>
+        <v-btn color="primary" @click=saveUsername>Save</v-btn>
+      </v-card-actions>
```
Comment (montalcy) [inline wordle-web/components/NameDialog.vue:11]: Is there a way to not have the username get typed as a guess in worlde? We had the same problem

File: wordle-web/pages/index.vue [hunk 10]
```diff
@@ -60,7 +67,7 @@ onUnmounted(() => {
 });
 
 async function getWordFromApi(): Promise<string> {
```
Comment (montalcy) [inline wordle-web/pages/index.vue:81]: Cool way to see this implemented, we haven't had a chance to get to this

File: wordle-web/pages/leaderboard.vue [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+<script setup lang="ts">
+
+import { ref } from 'vue';
+    import Axios from 'axios';
+
+const apiUrl = '';
+
+
+interface Player {
```
Comment (montalcy) [inline wordle-web/pages/leaderboard.vue]: I agree that the extra credit shouldnt be too bad to add in the seconds!

**Review by montalcy — COMMENTED** [review]
1. Create a leaderboard API Controller ✅

Create an endpoint that returns the top 10 scores (HttpGet) ✅
Results should include Name, average guesses, and number of games played ✅ make sure to format this neatly :) 
Create an endpoint that allows for posting a score (HttpPost) with data in the body ✅
A new name should add a new record  ❌ not necessarily implemented 
An existing name should update that record ✅
Save and load all data with Entity Framework to an Azure SQL database or LocalDb for debugging ✅❌ look like yall are working it out
Class should be named "Player" and "Players" for the DbSet ✅
Fields should be:
int PlayerId ✅
string Name ✅
int GameCount ✅
double AverageAttempts ❌
int AverageSecondsPerGame ❌(OPTIONAL)
The logic should reside in a service that is injected via dependency injection into the controller ✅
Set up CORS (Cross Origin Resource Sharing) ✅
2. Create a dialog for getting the user's name ✅

If we don't have the user's name, present the user with a dialog that allows them to type in their name ✅
The users current name should show on the game page (upper right) ❌ I didnt see it 
This should be visible when setting/changing the name. It should change as the user types ❌
The user should be able to click on their name to change it (with the same dialog) ❌
Save the name in local storage and load it for the next game if it is set ✅
If the name is not set, display Guest as the name. ✅
Clicking on this allows for preemptive setting of the name before the end of the game ✅
Dialog must be persistent, doesn't close by clicking off the dialog ❌
3. Save the Score and Seconds to complete ✅

Save the score and time to complete in seconds at the end of the game if we have a name. ✅
If the user hasn't set a name, save the score and seconds as "Guest". (Be sure to prompt first) ✅
Save should be done with an Axios Post ✅
4. Create a Leaderboard Page ✅

Create a /leaderboard page in the application that shows the top scores ✅
Use Axios to call the leaderboard API ✅
Make a nice display of the returned data ❌ no data displayed
Add a link on the front page to access the leaderboard ❌
Add a link to the app bar to access the leaderboard ❌
5. Unit testing ❌(OPTIONAL) (Not covered in class, but in instructor's repo: https://github.com/granterickson/EWU-CSCD379-2024-Spring)

Tests for the service ❌ (OPTIONAL)
Integration tests for the controller ❌ (OPTIONAL)
6. Deploy site to Azure ✅

Deploy API site with Github Action to Azure Web App ✅
Unit Tests should run on build ❌

File: Wordle.Api/Wordle.Api/Controllers/LeaderboardController.cs (hunk 1 above)
Comment (connerhanson) [reply Wordle.Api/Wordle.Api/Controllers/LeaderboardController.cs]: Refactored most of this code and fixed this thank you!

File: Wordle.Api/Wordle.Api/Services/WordleService.cs (hunk 4 above)
Comment (connerhanson) [reply Wordle.Api/Wordle.Api/Services/WordleService.cs:1]: Tests kept giving me weird errors that didn't make sense.

File: wordle-web/pages/leaderboard.vue (hunk 5 above)
Comment (connerhanson) [reply wordle-web/pages/leaderboard.vue]: Done!

File: .github/workflows/assignment3_agreeable-island-web-app.yml
Comment (connerhanson) [reply .github/workflows/assignment3_agreeable-island-web-app.yml:1]: You have too depending on what it is in each persons github secerets, the fork still had the one from megs machine.

File: Wordle.Api/Wordle.Api/Services/WordleService.cs (hunk 7 above)
Comment (connerhanson) [reply Wordle.Api/Wordle.Api/Services/WordleService.cs]: Possibly, though this implementation seems better to me since you don't have to make another method to check if the name is already in the database or not.

File: wordle-web/pages/leaderboard.vue (hunk 11 above)
Comment (connerhanson) [reply wordle-web/pages/leaderboard.vue]: Done!

File: wordle-web/app.vue (hunk 8 above)
Comment (connerhanson) [reply wordle-web/app.vue:8]: Its in the app bar, I though it was more pleasing to look at.

File: Wordle.Api/Wordle.Api/Dtos/PlayerDTO.cs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+namespace Wordle.Api.Dtos;
+
+public record PlayerDTO{
```
Comment (A-Tanner) [inline Wordle.Api/Wordle.Api/Dtos/PlayerDTO.cs:3]: It's a more common convention tor acronyms to be camel case and not fully capitalized (e.g. Dto instead of DTO)

File: Wordle.Api/Wordle.Api/Controllers/LeaderboardController.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    public async Task<PlayerDto> Post(PlayerDto request)
+    {
+        Player player = await _service.PostScoreAsync(request);
+        return new PlayerDto
+        {
+            Name = player.Name,
+            GameCount = player.GameCount,
+            AverageAttempts = player.AverageAttempts
+        };
+    }
```
Comment (A-Tanner) [inline Wordle.Api/Wordle.Api/Controllers/LeaderboardController.cs:34]: I'd recommend against this naming scheme for your controller methods. Your routes make very good method names, use those instead. A controller will often have multiple post and get methods. In this case, it works to just name the methods "Post" and "Get" since they are unqiue in that capacity, but naming conventions are important to observe in all scenarios

File: wordle-web/pages/index.vue [hunk 14]
```diff
@@ -1,72 +1,108 @@
...
+  else {
+    attempts = game.maxAttempts;
+  }
+  const requestBody = {
+    name: nuxtStorage.localStorage.get("userName", userName.value),
+    gameCount: 1, //since it posts after everygame
+    averageAttempts: attempts,
+    AverageSecondsPerGame: 0
+  };
+  axios.post("http://localhost:5297/Leaderboard", requestBody);
```
Comment (A-Tanner) [inline wordle-web/pages/index.vue:103]: When this app is deployed to Azure, it tries to communicate to 127.0.0.1:5297, but there is nothing there. This needs to target your API endpoint on Azure to save scores when you aren't running your API locally.

File: wordle-web/components/NameDialog.vue [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  <v-dialog v-model="modelValue" max-width="500">
+    <v-card>
+      <v-sheet color="primary"></v-sheet>
+      <v-card-title>Enter a username.</v-card-title>
+      <v-text-field v-model="userName" label="UserName" persistent></v-text-field>
+      <v-card-actions>
+        <v-spacer></v-spacer>
+        <v-btn color="primary" @click="closeDialog">Cancel</v-btn>
+        <v-btn color="primary" @click=saveUsername>Save</v-btn>
+      </v-card-actions>
```
Comment (A-Tanner) [inline wordle-web/components/NameDialog.vue:11]: Dialogue needs to be persisted. Here is the documentation for this attribute, you can look at the source code to see how it's accomplished

https://vuetifyjs.com/en/components/dialogs/#persistent

**Review by A-Tanner — COMMENTED** [review]
1. Create a leaderboard API Controller

Create an endpoint that returns the top 10 scores (HttpGet) ✅
Results should include Name, average guesses, and number of games played✅
Create an endpoint that allows for posting a score (HttpPost) with data in the body ✅
A new name should add a new record ✅
An existing name should update that record ✅
Save and load all data with Entity Framework to an Azure SQL database or LocalDb for debugging ✅
Class should be named "Player" and "Players" for the DbSet ✅
Fields should be:
int PlayerId ✅
string Name ✅
int GameCount ✅
double AverageAttempts ✅
int AverageSecondsPerGame ✅(OPTIONAL)
The logic should reside in a service that is injected via dependency injection into the controller ✅
Set up CORS (Cross Origin Resource Sharing) ✅
2. Create a dialog for getting the user's name ✅

If we don't have the user's name, present the user with a dialog that allows them to type in their name ✅
The users current name should show on the game page (upper right) ✅ -- No, the other right!

This should be visible when setting/changing the name. It should change as the user types ✅
The user should be able to click on their name to change it (with the same dialog) ✅
Save the name in local storage and load it for the next game if it is set ✅
If the name is not set, display Guest as the name. ✅
Clicking on this allows for preemptive setting of the name before the end of the game ✅
Dialog must be persistent, doesn't close by clicking off the dialog ❌
3. Save the Score and Seconds to complete ❌✅

Save the score and time to complete in seconds at the end of the game if we have a name. ❌ -- Time hardcoded to 0, post does not target API endpoint

If the user hasn't set a name, save the score and seconds as "Guest". (Be sure to prompt first) ❌ -- There is no prompt when completing a game as "Guest"

Save should be done with an Axios Post ✅
4. Create a Leaderboard Page ✅

Create a /leaderboard page in the application that shows the top scores ✅
Use Axios to call the leaderboard API ✅
Make a nice display of the returned data ✅
Add a link on the front page to access the leaderboard ✅
Add a link to the app bar to access the leaderboard ❌

5. Unit testing ❌ (OPTIONAL) (Not covered in class, but in instructor's repo: https://github.com/granterickson/EWU-CSCD379-2024-Spring)

Tests for the service ❌ (OPTIONAL)
Integration tests for the controller ❌ (OPTIONAL)
6. Deploy site to Azure ✅

Deploy API site with Github Action to Azure Web App ✅
Unit Tests should run on build ✅
Extra Credit
Unit testing on the client side (3) ❌
Animate the leaderboard (3) ❌
Create a logo (3) ❌
Add Average Seconds to win in score (3) ❌


Overall, strong submission that needs a bit more love and polish. Make sure you're saving time to respond to feedback in future assignments. While I can't diagnose the API issues you folks are experiencing based on the comments left, I'd be happy to pair with either or both of you over a Teams call some time to resolve this if it's still a problem. Shoot me a message!
